hydroflow_lang/src/graph/flat_graph_builder.rs

```diff
@@ -902,18 +902,42 @@ impl FlatGraphBuilder {
                 continue;
             }
             let op_span = operator.span();
+
             for (pred_edge, pred_node_id) in self
                 .flat_graph
                 .node_predecessors(node_id)
                 .collect::<Vec<_>>()
             {
-                if let Some(op_inst) = self.flat_graph.node_op_inst(pred_node_id) {
-                    if "persist" == op_inst.op_constraints.name
-                        || "state" == op_inst.op_constraints.name
-                    {
-                        continue;
+                // Check if the predecessor is already persisted. If so, no persist should be inserted.
+                let mut pred_persisted = true;
+                {
+                    let mut pred_node_ids = VecDeque::from_iter([pred_node_id]);
+                    while let Some(pred_node_id) = pred_node_ids.pop_front() {
+                        let Some(op_inst) = self.flat_graph.node_op_inst(pred_node_id) else {
+                            pred_persisted = false;
+                            break;
+                        };
+                        match op_inst.op_constraints.name {
+                            // Stateful operators, we're good.
+                            "persist" | "state" => {}
+                            // Identity operators, check transitively.
+                            // TODO(mingwei): Add `transparent` or `is_identity` field to
+                            // `OperatorInstance` so we don't have to list all variants here.
+                            "identity" | "union" | "tee" => {
+                                pred_node_ids
+                                    .extend(self.flat_graph.node_predecessor_nodes(pred_node_id));
```

Comment:
Should we turn the responsibility around to the operators? The operator should tell this part of the code what interests it.  

In other words. Should there be an `operator.persisted_at_end()` or something of that nature, this code only checks if that is true.

Reply:
Yeah had the TODO for it, I will make it a separate issue

Reply:
https://github.com/hydro-project/hydroflow/issues/1311